designs/node-ownership.md

```diff
@@ -0,0 +1,227 @@
+
+1. Capture in-flight capacity in a way so that we can quickly complete Karpenter’s provisioning loops and continue to start the next round of provisioning while having a consistent view of cluster state
+2. Completely remove (or, at a minimum, limit) the mutation of the Node so that we don’'t alter the Node in ways that the user doesn't expect
+3. Have little to no impact on cluster performance, including memory usage in etcd and query performance on the kube-apiserver.
+
+Fundamentally, if we do not capture in-flight capacity in the Node object, we need some other place to capture capacity so that we can fulfill #1 listed above. This necessitates some kind of “store” to keep in-flight node data, whether this is an in-memory store or the Kubernetes api-server (etcd) store. The two options presented below highlight both options, including their tradeoffs related to the fundamentals mentioned above.
+
+### Kubernetes CRD Object Store (Machine CRD)
+
+Karpenter will no longer create node objects or launch instances as part of the provisioning loop, but, instead, will create Machine CRs at the completion of the provisioning loop. This machine CR will then be picked up by a separate controller that will launch capacity based on the requirements passed from the provisioning loop and will resolve the static values from the **CreateFleet** response into its status. After the instance is launched, the kubelet starts, and the node joins the cluster, machines will be mapped to nodes using the `spec.providerID` of the Node and the `status.providerID` of the Machine.
+
+#### Impact
+
+1. Introduction of the Machine CR as a representation of a “scheduling decision” to launch by the CloudProvider
+2. Migration of Deprovisioning mechanisms from watching and monitoring the Node object to watching and monitoring the Machine CR
+3. Migration of the Node termination finalizer to Machine to orchestrate the cordoning, draining, and deletion
+4. Required Garbage Collection of any Machines that don’t have instances mapped to them
+5. Liveness checks for Nodes that don’t register or become ready after a certain TTL to delete the backing Machine
+6. In-flight representations of Nodes for scheduling become the Machine `.status.allocatable` until the Node registers and initializes its `.status.allocatable`
+7. Mirroring Machine labels and annotations onto Nodes through a reconciliation mechanism after Node join
+
+### In-Memory Node Store
+
+An alternative to storing persistent data on the Kubernetes API server is to utilize the cloud provider’s instance data from a **ListMachines** call and utilize the instance data that is stored in cloud provider’s instance store to understand our in-flight capacity. This remote store means we no longer need to create the node for inflight capacity, but we still need to operate on the node object for details like ownership and node termination/provisioner ownership.
```

Comment:
Another alternative: store pending instance information in the `.status` of each Provisioner, and limit the number of in-flight VM launches.

Analogy: the Job API in Kubernetes uses `.status` to track completed Jobs prior to being able to track them, making atomic updates to the Job to record a count of completions.

Reply:
This approach also gives Karpenter a way to recover from an outage that occurs after creating the Fleet and before recording the new instance: on detecting a suitably tagged fleet, Karpenter can record that detection in the provisioner's `.status`.

This still might need a CR that is specific to the particular cloud provider (like an AWSNodeTemplate, but for provider-specific status). I'm not sure.

Reply:
I think this still relies on a reconciliation mechanism to LIST out the instances that have already been created based on tag in the case of failure to store into the status like you have described.

This, to me, seems like an extension of keeping track of details in-memory (notably, a more observable one) since we will still have to perform essentially a LIST/WATCH in order to deal with crashes/restarts/outages.

Reply:
Storing status into the cluster's API is _not_ the same as keeping it withing Karpenter's memory. In the first case, a cluster operator can observe Karpenter's view of the actual state.